Sampling heap profiler

StephanDollberg opened the pull request:
As discussed this is the current state:

Commits should stand on their own. The most interesting ones are probably "use sampler" and "use separate pool" which implement the meat. 

It's not optimized yet to the last instruction as that will require reshuffling a bit of code but I first want to confirm we are happy with the general approach for a v1.

I am using this with https://github.com/redpanda-data/redpanda/compare/dev...StephanDollberg:redpanda:support-mem-profiling?expand=1 (seastar change not committed there):



I am using OMB benchrunner using this config: https://gist.github.com/StephanDollberg/3d51e3ea029708f7aed49a277220cb26 I am happy to try other ones if you think they are more fitting.

These are the last lines of three different runs each:

current dev:

```
[client 0:0] 16:45:01.158 [main] INFO - E2E Latency (ms) avg: 3.828 - 50%: 3.731 - 99%: 7.073 - 99.9%: 11.093 - Max: 19.623
[client 0:0] 16:45:01.413 [main] INFO - ----- Aggregated Pub Latency (ms) avg: 2.689 - 50%: 2.253 - 95%: 3.460 - 99%: 6.235 - 99.9%: 115.282 - 99.99%: 297.557 - Max: 566.035 | Pub Delay (us)  avg: 169.938 - 50%: 24.000 - 95%: 64.000 - 99%: 77.000 - 99.9%: 57342.000 - 99.99%: 68196.000 - Max: 72037.000

[client 0:0] 16:40:11.059 [main] INFO - E2E Latency (ms) avg: 3.922 - 50%: 3.798 - 99%: 7.706 - 99.9%: 11.915 - Max: 23.610
[client 0:0] 16:40:11.302 [main] INFO - ----- Aggregated Pub Latency (ms) avg: 2.711 - 50%: 2.288 - 95%: 3.703 - 99%: 6.601 - 99.9%: 101.424 - 99.99%: 250.523 - Max: 468.353 | Pub Delay (us)  avg: 157.477 - 50%: 25.000 - 95%: 64.000 - 99%: 77.000 - 99.9%: 53500.000 - 99.99%: 67445.000 - Max: 74282.000

[client 0:0] 16:29:43.995 [main] INFO - E2E Latency (ms) avg: 4.033 - 50%: 3.884 - 99%: 7.773 - 99.9%: 11.462 - Max: 18.480
[client 0:0] 16:29:44.260 [main] INFO - ----- Aggregated Pub Latency (ms) avg: 3.001 - 50%: 2.337 - 95%: 3.967 - 99%: 7.399 - 99.9%: 170.877 - 99.99%: 398.539 - Max: 609.247 | Pub Delay (us)  avg: 219.166 - 50%: 24.000 - 95%: 64.000 - 99%: 77.000 - 99.9%: 79371.000 - 99.99%: 96960.000 - Max: 103688.000
```

1MB sampling:

```
[client 0:0] 17:04:58.938 [main] INFO - E2E Latency (ms) avg: 4.207 - 50%: 4.049 - 99%: 7.960 - 99.9%: 12.608 - Max: 26.739
[client 0:0] 17:04:59.205 [main] INFO - ----- Aggregated Pub Latency (ms) avg: 3.125 - 50%: 2.433 - 95%: 4.011 - 99%: 6.609 - 99.9%: 186.264 - 99.99%: 369.935 - Max: 591.547 | Pub Delay (us)  avg: 216.450 - 50%: 24.000 - 95%: 64.000 - 99%: 78.000 - 99.9%: 71178.000 - 99.99%: 97655.000 - Max: 99394.000

[client 0:0] 17:07:49.009 [main] INFO - E2E Latency (ms) avg: 4.190 - 50%: 4.027 - 99%: 8.137 - 99.9%: 12.315 - Max: 32.772
[client 0:0] 17:07:49.294 [main] INFO - ----- Aggregated Pub Latency (ms) avg: 3.164 - 50%: 2.464 - 95%: 4.232 - 99%: 7.424 - 99.9%: 173.275 - 99.99%: 417.215 - Max: 657.475 | Pub Delay (us)  avg: 312.114 - 50%: 24.000 - 95%: 64.000 - 99%: 80.000 - 99.9%: 99721.000 - 99.99%: 127613.000 - Max: 130623.000

[client 0:0] 17:12:33.201 [main] INFO - E2E Latency (ms) avg: 3.968 - 50%: 3.855 - 99%: 7.483 - 99.9%: 11.334 - Max: 18.778
[client 0:0] 17:12:33.452 [main] INFO - ----- Aggregated Pub Latency (ms) avg: 2.799 - 50%: 2.363 - 95%: 3.746 - 99%: 6.397 - 99.9%: 118.937 - 99.99%: 279.579 - Max: 449.807 | Pub Delay (us)  avg: 135.455 - 50%: 24.000 - 95%: 64.000 - 99%: 76.000 - 99.9%: 55372.000 - 99.99%: 79435.000 - Max: 85011.000
```

So it seems like it maybe has a ~2% perf impact but it certainly is in the noise. I am not sure whether we have nice ways of aggregating these or comparing them.

The following is a flamegraph generated during one such run (with the periodic logging printing the top-20 stacks):

![flamegraph](https://user-images.githubusercontent.com/2904723/234948764-db9b3e03-c5ba-4f9d-a934-8926d4932f4b.svg)

Note the output-to-flamegraph script currently doesn't strip return types from functions so it looks a bit meh.

I will work on exposing the fullset from the admin API next.

StephanDollberg commented:
Today I did a run with 64k sampling frequency:

```
[client 0:0] 10:45:36.295 [main] INFO - E2E Latency (ms) avg: 4.489 - 50%: 4.317 - 99%: 8.796 - 99.9%: 14.399 - Max: 33.170
[client 0:0] 10:45:36.557 [main] INFO - ----- Aggregated Pub Latency (ms) avg: 3.315 - 50%: 2.741 - 95%: 4.749 - 99%: 7.724 - 99.9%: 139.540 - 99.99%: 307.537 - Max: 488.957 | Pub Delay (us)  avg: 193.453 - 50%: 24.000 - 95%: 64.000 - 99%: 77.000 - 99.9%: 67339.000 - 99.99%: 93158.000 - Max: 96961.000k
```

Impact is more clearly visible there.

I also implemented the retrieval of the full stack in the Kafka API. Below is a full sample with 1MB sampling:

![during_flamegraph](https://user-images.githubusercontent.com/2904723/235165519-6347d72d-a281-47e3-a185-5e03e2999d71.svg)

I will probably try to apply some hacky filtering for the return types in the script. But will also need to read up, maybe one can avoid that with addr2line or llvm-symbolize or something.

StephanDollberg commented on src/core/memory.cc:
```diff
@@ -245,6 +246,76 @@ static uint64_t get(types stat_type)
+// would require storing the weight in addition to the alloation site ptr as on
+// free we need to know how much a sample accounted for. Hence, for now we
+// simple always use the sample frequency.
+class Sampler {
```
> The original source code is under an Apache 2.0 license. Does this license allow for copying subsets of the work with attribution?

I am fairly certain but will double check.

 > We have a goal of up-streaming this PR to scylladb/seastar. Do they accept code under an Apache 2.0 license? And if so do they accept code that isn't formatted according to their styling guide?

I was worried at first as well but then saw that seastar is actually also apache2 licensed so I hope this shouldn't be a problem.

Re formatting, I just dumped the whole thing in for now but I will probably take it apart a bit so that we can safe a couple instructions at which point the code will probably not really resemble the original anymore and I will be able to format as per seastar standard (not saying I would want to remove the mention of the original one).

StephanDollberg commented on include/seastar/core/memory.hh:
```diff
@@ -414,10 +414,10 @@ std::vector<allocation_site> sampled_memory_profile();
 /// or a zoomable flame graph ([flame graph generation instructions](https://github.com/scylladb/scylla/wiki/Seastar-heap-profiler),
 /// [example flame graph](https://user-images.githubusercontent.com/1389273/72920437-f0cf8a80-3d51-11ea-92f0-f3dbeb698871.png)).
-void set_heap_profiling_enabled(bool);
+void set_heap_profiling_enabled(size_t);
```
Yep sounds good.

StephanDollberg commented on src/core/memory.cc:
```diff
@@ -869,13 +948,13 @@ small_pool::alloc_site_holder(void* ptr) {
 
 void*
-cpu_pages::allocate_small(unsigned size) {
+cpu_pages::allocate_small(unsigned size, size_t sample_weight) {
```
Sorry yeah that's not really clean right now. Something I will clean up.

I started with the bool, then when introducing the proper sampler switched to the weight and then again realised that we can't actually use the weight right now because we would have to store it so ended up using it like a bool.

StephanDollberg commented on src/core/memory.cc:
```diff
@@ -574,7 +577,8 @@ struct cpu_pages {
     static constexpr unsigned nr_span_lists = 32;
     page_list free_spans[nr_span_lists];  // contains aligned spans with span_size == 2^idx
-    small_pool_array small_pools;
+    small_pool_array<false> small_pools;
```
Might do some more thorough testing but my napkin math agreed with yours. Hence, I agree it shouldn't be much of an issue.

Also because of the sampling, for example assuming 1MB sampling and actually sampling one of these allocations it's very likely you will actually sample more allocations of the same and actually use some of that wastage.

StephanDollberg commented on scripts/symbolize_mem_dump.py:
```diff
@@ -0,0 +1,95 @@
+        if 'current_backtrace_tasklocal' in line:
+            continue
+
+        if line == 'seastar::memory::get_backtrace()':
```
Some version thereof is in the gdb script, yes.

StephanDollberg commented on src/core/memory.cc:
```diff
@@ -155,31 +156,13 @@ __thread volatile int critical_alloc_section = 0;
-struct hash<seastar::allocation_site> {
-    size_t operator()(const seastar::allocation_site& bi) const {
-        return std::hash<seastar::saved_backtrace>()(bi.backtrace);
+struct hash<seastar::memory::allocation_site> {
```
No not really. I originally wasn't sure whether the `allocation_site` struct would be a good fit as is.

I can't see anything wrong with it so we can move this over as well (and I can also add some comments).

StephanDollberg commented on src/core/memory.cc:
```diff
@@ -1398,12 +1541,13 @@ void* allocate(size_t size) {
     if (size <= sizeof(free_object)) {
         size = sizeof(free_object);
     }
+    size_t sample_weight = get_cpu_mem().should_sample(size);
```
You are definitely correct. Not entirely sure I understand the location of the comment though as this is in `allocate` and not `free`. 

I am wondering whether github tricked you here with the diff view? 

<img width="655" alt="image" src="https://user-images.githubusercontent.com/2904723/235702097-0f868c9f-6e9d-44ca-a293-254fb7a34b9a.png">

StephanDollberg commented:
Thanks both @travisdowns / @ballard26 

From the comments I understand we don't see any fundamental issues with approach taken. 

Hence, as next steps:

 - I will write this down in the RFC
 - Clean this up a little bit more and then open for review/merge. 

Unless you have any preference I will just work on top of the current set of commits. Then you guys don't have to re-review everything.

StephanDollberg commented:
I have amended a few commits which address the cleaning up I still wanted to do and hopefully also some of the comments you guys had above.

The commits should stand on their own so reviewing commit by commit should be fine if you prefer that over the whole thing. If you prefer I can also do a rebase now and integrate the new commits into the old ones (instead of doing it later).

There isn't any major functional change, the areas of change are:
 - General clean up (more comments, sample_weight vs. should_sample). If you think anything is still undercommented please just let me know and I will expand.
 - Optimize the allocation and sampler a bit, we have less duplicate "should sample" checks now and the fast path in the sampler is also simpler.
 - I removed the python script. I am not entirely happy with it yet and there is no need to merge it now as it's not bound to a release
 - I am inclined to move the sampler to a different file as that potentially might make licensing questions easier and also rebasing if for example seastar people would want a dumb "sample every n" sampler instead

As mentioned further above, if you guys want me to run any specific benchmarks then just let me know which (spec). I will tomorrow try to get a better summary out of plot.py.

I will also shortly open the RP PR.

StephanDollberg commented:
Moved the sampler to its own file and a note to the NOTICE file which in my understanding should answer the licensing questions. 

Let me know if you feel differently.

StephanDollberg commented:
As discussed, have moved the sampler changes into two separate commits (first adding, second our modifications)

StephanDollberg commented on include/seastar/core/memory.hh:
```diff
@@ -382,29 +383,76 @@ public:
 /// This script can generate either textual representation of the data,
 /// or a zoomable flame graph ([flame graph generation instructions](https://github.com/scylladb/scylla/wiki/Seastar-heap-profiler),
 /// [example flame graph](https://user-images.githubusercontent.com/1389273/72920437-f0cf8a80-3d51-11ea-92f0-f3dbeb698871.png)).
-void set_heap_profiling_enabled(bool);
```
> So this "replaces" the old sampling mode of sample-everything to be inspected in gdb, right?

The gdb still works

 > This might not fly upstream, we'll have to see. Maybe let's focus on the RP side of things for now.

I very much agree with this. IMO I'd prefer to just remove the GDB way (as it means we could get rid of those extra allocation site links) but at the end of the day upstream will decide and I don't think there is much point in guess in now.

StephanDollberg commented on include/seastar/core/memory.hh:
```diff
@@ -382,29 +383,76 @@ public:
+void set_heap_profiling_sampling_rate(size_t);
+
+/// Checks whether heap profiling is currently enabled
+bool get_heap_profiling_enabled();
```
Brandon suggested to have a boolean version further above. I am happy to add another function.

Though maybe it's in the category of "we don't need it, lets see what seastar people want"?

ballard26 replied:
My thoughts on it are that `sampling_rate` and `enabled` are functionally independent configuration options. From that one should be able to change without effecting the other. And it's what I'm doing in the CPU profiler with a set of `get/set` methods for each.

That being said I'm not dead set on it if. And there is some sense to just have `sampling_rate == 0` entail that the profiler is disabled.

StephanDollberg replied:
Fair, have changed it back.

StephanDollberg commented on include/seastar/core/memory.hh:
```diff
@@ -382,29 +383,76 @@ public:
-/// Enable/disable heap profiling.
+/// Describes an allocation location in the code.  The location is identified by
+/// its backtrace. One allocation_site can represent many allocations at the
+/// same location. `count` and `size` represent the cumulative sum of all
```
Good point, I will clarify that. Right now it is infact the grossed up value (see sampler.sample_size() and its usage). I.e.: we don't store the original size but the how-much-it-should-count for size directly. So no post processing is needed.

We could obviously store the original size but I currently don't see an advantage of doing that given that making use of that would require additional information (storing per allocation information) and it has the benefit that in postprocessing you don't need the sampling rate.

StephanDollberg commented on include/seastar/core/memory.hh:
```diff
@@ -382,29 +383,76 @@ public:
+/// Copies the current sampled set of allocation_sites in the output parameter
+/// (up to size of output vector) Returns amount of copied elements. Useful if
+/// one wants to avoid allocating an output vector (e.g.: under OOM conditions)
+size_t sampled_memory_profile(std::vector<allocation_site>&);
```
Good idea, though I guess you lock your API into a certain interface then.

Could make it a template I guess but that would mean exposing/moving a lot of the memory.cc private stuff into the header.

StephanDollberg commented on include/seastar/core/memory.hh:
```diff
@@ -382,29 +383,76 @@ public:
 /// For more information about heap profiling see
-/// \ref set_heap_profiling_enabled().
+/// \ref set_heap_profiling_sampling_rate().
 class scoped_heap_profiling {
```
> How does nesting work here?

It doesn't really work with mixed sampling rates. I will make a note of that.

Please see the super comment below for an explanation.

StephanDollberg commented on include/seastar/util/sampler.hh:
```diff
@@ -0,0 +1,128 @@
+//  multiple loops of drawing from the exp distribution). The reason is that
+//  in our memory sampler we would have to store the weight in addition to the
+//  alloation site ptr as on free we need to know how much a sample accounted
+//  for. Hence, for now we simply always use the sampling interval.
```
Yes, in post-processing or as it's currently done in the size tracking directly.

StephanDollberg commented:
Couple comments above touched on this so thought I put a summary and some background on changing sampling size, scoped profiling etc:

Fundamentally changing the sampling frequency doesn't really work right now. The reason is that for that we'd need to track the sample rate/sample weight per allocation so that on free we can reduce the allocation site size by the correct amount (given we store the weighted size and not the original raw size). Note this doesn't really change if you store the original size as then when taking a snapshot you'd have to collect each individual allocation plus its weighted size. 

With sampling I am not sure how useful the scoped profiling still is. I guess it was for cases where one didn't want to impose the overhead on everything during an investigation and only look at a small bit but I feel like sampling should solve such usecases. As mentioned above, scoped does still work just changing sample rates in between doesn't really as per above reasons.

I think this is another area where we will just have to see what seastar people think/want. They reasons above are solvable but require more overhead so I don't think fixing them is worth it without a concrete usecase.

StephanDollberg commented:
As discussed yesterday, I did run a longer 750MB / 50P / 50C OMB run today. Two runs each, one with latest dev build and another with dev+memory sampler PR.

Below, e2e latency distribution, p99 pub latency and p50 e2e latency. I think if anything it shows that OMB is too noisy to detect smaller changes. 

<img width="1131" alt="image" src="https://github.com/redpanda-data/seastar/assets/2904723/48ed7054-4818-44da-801a-249fb947bda7">
<img width="1162" alt="image" src="https://github.com/redpanda-data/seastar/assets/2904723/1957cbe2-1cf5-4732-9c02-12870ebf7390">
<img width="1160" alt="image" src="https://github.com/redpanda-data/seastar/assets/2904723/6acb66fd-0c1b-46ac-b4ce-b220fca5e1ae">

StephanDollberg commented:
Just posting the microbenchmark results here again for clarity as they are now hidden int the commits:

```
v23.2.x

test                                      iterations      median         mad         min         max      allocs       tasks        inst
allocator.single_alloc_and_free_small      270469115     3.701ns     0.009ns     3.692ns     3.712ns       1.000       0.000       100.0
allocator_test.single_alloc_and_free_small_many    18155337    54.877ns     0.034ns    54.798ns    54.918ns      10.000       0.000       985.0
allocator_test.single_alloc_and_free_small_many_cross_page    47225352    21.082ns     0.017ns    21.049ns    21.116ns       5.000       0.000       491.0
allocator_test.single_alloc_and_free_small_many_cross_page_alloc_more     1036909   961.142ns     1.394ns   959.643ns   990.267ns     101.000       0.000     16881.0
allocator.single_alloc_and_free_large       82006844    11.830ns     0.002ns    11.823ns    11.840ns       1.000       0.000       348.0

sampler

test                                      iterations      median         mad         min         max      allocs       tasks        inst
allocator.single_alloc_and_free_small      232397269     4.302ns     0.001ns     4.298ns     4.303ns       1.000       0.000       112.0
allocator_test.single_alloc_and_free_small_many    19644117    50.888ns     0.001ns    50.802ns    50.889ns      10.000       0.000      1105.0
allocator_test.single_alloc_and_free_small_many_cross_page    46667620    21.507ns     0.006ns    21.501ns    21.593ns       5.000       0.000       551.0
allocator_test.single_alloc_and_free_small_many_cross_page_alloc_more     1065955   942.539ns     0.404ns   942.090ns   943.257ns     101.000       0.000     17907.0
allocator.single_alloc_and_free_large       79390976    12.607ns     0.002ns    12.590ns    12.610ns       1.000       0.000       392.0
```

Those numbers are fairly stable. Two things to note.

First is that these numbers are with sampling compiled in but not enabled. The fast path (i.e.: we are not taking a sample) in the current implementation is optimized for having the sampler turned on so there is actually no difference between having it turned on or off. We know that taking a sample is expensive because of the backtrace, equally microbenchmarking with sampling turned on is difficult because of the probabilistic nature, varying iterations and warmup etc. 

Another thing to note, we could bring down the large_allocation overhead and instruction count by force inlining some of the intermediate large allocation layers. I haven't done this here because Travis mentioned his patches do something like that anyway.

StephanDollberg commented:
> Is it possible to turn sampling on but set the period to be so large that it never samples?

Yes, then you get the same numbers. As it's basically identical to setting _interval_to_next_sample to INT64_MAX which is what we do if turned off.

(single_alloc_and_free_small_many_cross_page_alloc_more perf test was actually showing a differnce because we had a leftover disable_backtrace_temporarily but that's fixed now)

StephanDollberg commented:
I have incorporated Travis' optimizations now. 

Updated numbers are:

```
v23.2.x

test                                      iterations      median         mad         min         max      allocs       tasks        inst
allocator_test_split.alloc_only                                         198864000     3.292ns     0.001ns     3.290ns     3.293ns       1.000       0.000        55.3
allocator_test_split.free_only                                          201541000     1.664ns     0.000ns     1.664ns     1.665ns       0.000       0.000        45.3
allocator_test_split.alloc_free                                         188049000     4.933ns     0.001ns     4.926ns     4.934ns       1.000       0.000       100.5
allocator_test.single_alloc_and_free_small_many                          19280658    51.687ns     0.055ns    51.534ns    51.873ns      10.000       0.000       965.0
allocator_test.single_alloc_and_free_small_many_cross_page               49127098    20.365ns     0.013ns    20.343ns    20.378ns       5.000       0.000       481.0
allocator_test.single_alloc_and_free_small_many_cross_page_alloc_more     1029357   973.258ns     0.515ns   972.743ns   974.655ns     101.000       0.000     16471.0
allocator_test_split.alloc_only_large                                    67932000     9.945ns     0.002ns     9.935ns     9.947ns       1.000       0.000       219.3
allocator_test_split.free_only_large                                     67644000     4.975ns     0.003ns     4.973ns     4.984ns       0.000       0.000       134.4
allocator_test_split.alloc_free_large                                    65357000    15.065ns     0.010ns    15.023ns    15.075ns       1.000       0.000       353.7

sample before travis:

allocator_test_split.alloc_only                                         240979000     2.383ns     0.001ns     2.382ns     2.385ns       1.000       0.000        66.2
allocator_test_split.free_only                                          245268000     1.684ns     0.000ns     1.683ns     1.685ns       0.000       0.000        46.2
allocator_test_split.alloc_free                                         224929000     4.060ns     0.002ns     4.058ns     4.063ns       1.000       0.000       112.5
allocator_test.single_alloc_and_free_small_many                          18699835    53.441ns     0.050ns    53.391ns    53.559ns      10.000       0.000      1085.0
allocator_test.single_alloc_and_free_small_many_cross_page               45362372    22.317ns     0.016ns    22.301ns    22.362ns       5.000       0.000       541.0
allocator_test.single_alloc_and_free_small_many_cross_page_alloc_more     1052954   953.547ns     0.350ns   952.760ns   954.353ns     101.000       0.000     17608.0
allocator_test_split.alloc_only_large                                    54939000    11.121ns     0.001ns    11.118ns    11.123ns       1.000       0.000       244.3
allocator_test_split.free_only_large                                     55047000     5.639ns     0.000ns     5.635ns     5.643ns       0.000       0.000       150.7
allocator_test_split.alloc_free_large                                    54017000    16.715ns     0.003ns    16.710ns    16.729ns       1.000       0.000       395.0


sample after travis:

allocator_test_split.alloc_only                                         273779000     2.197ns     0.000ns     2.196ns     2.198ns       1.000       0.000        57.2
allocator_test_split.free_only                                          274242000     1.445ns     0.000ns     1.444ns     1.458ns       0.000       0.000        37.2
allocator_test_split.alloc_free                                         246644000     3.683ns     0.001ns     3.683ns     3.685ns       1.000       0.000        94.5
allocator_test.single_alloc_and_free_small_many                          28357912    35.183ns     0.076ns    34.821ns    35.259ns      10.000       0.000       905.0
allocator_test.single_alloc_and_free_small_many_cross_page               62277115    16.042ns     0.008ns    16.034ns    16.059ns       5.000       0.000       451.0
allocator_test.single_alloc_and_free_small_many_cross_page_alloc_more     1111279   901.666ns     0.204ns   900.931ns   902.413ns     101.000       0.000     15765.0
allocator_test_split.alloc_only_large                                    58402000    11.401ns     0.004ns    11.396ns    11.409ns       1.000       0.000       250.3
allocator_test_split.free_only_large                                     58884000     5.516ns     0.001ns     5.512ns     5.518ns       0.000       0.000       143.7
allocator_test_split.alloc_free_large                                    57334000    16.958ns     0.002ns    16.947ns    16.960ns       1.000       0.000       394.0

```

I will now do a proper rebase and squash.

StephanDollberg commented:
Rebased and squashed, this should be good to go.

StephanDollberg commented:
Noticed that the cpp20 builds should actually all pass. Applied some minor fixes for gcc (rename and apply inline tag to all always_inline functions)
